File: kernel_design/executable_spec/kernel/worker_result_runtime.py
from __future__ import annotations

from abc import ABC, abstractmethod
from typing import Sequence

from .worker_delivery import (
    DeliveryReport,
    DeliveryReportOutcomeClass,
)


class WorkerResultRuntime(ABC):
    """Best-effort DeliveryReport queues partitioned by outcome class."""

    @abstractmethod
    def append_worker_results(
        self,
        *,
        results: Sequence[DeliveryReport],
    ) -> int:
        pass

    @abstractmethod
    def consume_worker_results(
        self,
        *,
        outcome_class: DeliveryReportOutcomeClass,
        limit: int,
    ) -> tuple[DeliveryReport, ...]:
        pass

File: kernel_jvm/src/main/java/com/xa/mass/kernel/delivery/WorkerResultRuntime.java
package com.xa.mass.kernel.delivery;

import com.xa.mass.workerdelivery.protocol.WorkerDeliveryProtocol.DeliveryReport;
import com.xa.mass.workerdelivery.protocol.WorkerDeliveryProtocol.DeliveryReportOutcomeClass;
import java.util.List;

public interface WorkerResultRuntime {

    int appendWorkerResults(List<DeliveryReport> results);

    List<DeliveryReport> consumeWorkerResults(
            DeliveryReportOutcomeClass outcomeClass,
            int limit
    );
}
